bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java

```diff
@@ -259,6 +265,20 @@ private void executeEventCommands(List<Event> events, CommandTagType execTime) {
         }
     }
 
+    /**
+     * Migration for last_update-channel as this change is compatible to previous instances.
+     */
+    private void migrateLastUpdateChannel() {
+        Thing thing = getThing();
+        if (thing.getChannel(CHANNEL_LAST_UPDATE) == null) {
+            logger.debug("last_update channel is missing in this Thing. Adding it.");
+            ThingBuilder thingBuilder = editThing();
+            ChannelBuilder channelBuilder = ChannelBuilder.create(new ChannelUID(thing.getUID(), CHANNEL_LAST_UPDATE));
```

Comment:
Please use the handler callback to create new channels (see [createChannelBuilder ()](https://github.com/openhab/openhab-core/blob/af266b5d9f0076deeb034ec5b67fd664a09f85e0/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/ThingHandlerCallback.java#L118-L126)).

Reply:
I know this method (and use it in event filter). However, at the time the migration gets called, the handler is not initialised yet. I did not want to run the migration on every update, if possible. Should i do instead?

Reply:
There is no need for the handler to be initialized to use this method. The in general is available when a handler is available.

The problem with the current implementation is, that you creates a "blank" channel referring to a channel type uid. No label, no description, no translation, no state description, and so on. The mentioned method requests all registered channel type registries to predefine all known (and translated) information for the channel to be created.

It should be the recommended way to create Channels.

Reply:
I'm referring with "has to be initialised" in this case to the [`getCallback()`](https://openhab.org/javadoc/latest/org/openhab/core/thing/binding/basethinghandler#getCallback()) call which is required to retrieve the `ThingHandlerCallback`. Is there a different way to retrieve it?

Or should i move the addition of this channel to the regular update of channels when already online?

Reply:
Hm. Looking into the implementation of `EventFilterHandler` i've done the update also while initializing the thing. It seems i thought "Only returns `null` while the handler is not initialized" differently than it is meant.

I'll change that in a separate MR as this one is already merged.

Reply:
Hm. Looking into the implementation of `EventFilterHandler` i've done the update also while initializing the thing. It seems i thought "Only returns `null` while the handler is not initialized" differently than it is meant.

I'll change that in a separate MR as this one is already merged.